### DISCUSSION TITLE
Do we promote flatpak or host based development?

### DISCUSSION THREAD

**Discussion by Antolius** [elementary/docs#195] [General]
Developer docs currently strike a balance between describing host based development flow (where dependencies and build tools are all installed locally and apps are built directly with meson) and flatpak based flow (where apps are built inside the flatpak container and the host machine doesn't have required libraries / tooling). Having to document both approaches can be tedious, and in interest of keeping docs concise we sometimes sacrifice details for one of the flows.

Would it make sense to commit to one approach and stick with it for most if not all docs sections?

Gnome builder went the flatpak route. That makes dependency management easier which is more important in gnome's case as they are a DE, not a distro. With elementary we have more control over what is available from official repository. Building apps inside flatpak container makes certain things more complicated (like checking gsettings, debugging, etc.). On the other hand apps are eventually going to be distributed as flatpaks on the AppCenter, and running flatpak versions during development helps detect ans fix container related issues early.

What do you think?

**Comment by danirabbit** [comment]
Yeah I think we should probably go all in on Flatpak based, especially as the platform can be out of sync with the host in terms of newer Gtk features, etc and some things are just different when you test them sandboxed versus in host

**Reply by Antolius** [reply]
Makes sense, dependency management is a big benefit.

Perhaps we could sprinkle some more tips about working with flatpak through the docs. Or maybe organize it as its own section and link to it? I think it depends on where the reader is coming from: if they are used to local host development and don't know how to do a certain thing within a container then a dedicated section seems more appropriate. If they are a new developer then they don't care about flatpak as a technology as much as they want to accomplish a specific task. In this case it makes sense to have the info embedded in the existing section.

Some flatpak specific topics that come to mind:
- Setting environment variables. This would be useful info for the Creating Logs section.
- Working with GSettings, basically an alternative to using dconf editor on the host.
- Finding various files created inside the container. This can be useful for troubleshooting, for example to verify that during installation meson put all the required files in the correct places. Or opening files created by the app with host tools, for example an sqlite db file.
- Generating files for translations. I must admit I don't know how to do this. 😅 Perhaps we could teach devs how to set up automated CI and do it there, like for 1st party apps? In any case we can update the Translations docs section with the new steps.

Perhaps there are other developer activities that we don't discuss currently which could be written from flatpak perspective. For example automated testing (discussed in #167), or debugging.
